text_collector_examples/smartmon.sh

```diff
@@ -0,0 +1,139 @@
+
+parse_smartctl_attributes() {
+  local disk="$1"
+  local disk_type="$2"
+  local labels="disk=\"${disk}\",type=\"${disk_type}\""
+  local vars="$(echo "${smartmon_attrs}" | xargs | tr ' ' '|')"
+  sed 's/^ \+//g' \
+    | awk -v labels="${labels}" "${parse_smartctl_attributes_awk}" 2>/dev/null \
+    | tr A-Z a-z \
+    | grep -E "(${smartmon_attrs})"
+}
+
+parse_smartctl_info() {
+  local -i smart_available=0 smart_enabled=0
+  local disk="$1" disk_type="$2"
+  while read line ; do
+    info_type="$(echo "${line}" | cut -f1 -d: | tr ' ' '_')"
+    info_value="$(echo "${line}" | cut -f2- -d: | sed 's/^ \+//g')"
+    case "${info_type}" in
+      Model_Family) model_family="${info_value}" ;;
+      Device_Model) device_model="${info_value}" ;;
+      Serial_Number) serial_number="${info_value}" ;;
+      Firmware_Version) fw_version="${info_value}" ;;
+      Vendor) vendor="${info_value}" ;;
+      Product) product="${info_value}" ;;
+      Revision) revision="${info_value}" ;;
+      Logical_Unit_id) lun_id="${info_value}" ;;
+    esac
+    if [[ "${info_type}" == 'SMART_support_is' ]] ; then
+      case "${info_value:0:7}" in
+        Enabled) smart_enabled=1 ;;
+        Availab) smart_available=1 ;;
+        Unavail) smart_available=0 ;;
+      esac
+    fi
+  done
+  if [[ -n "${model_family}" ]] ; then
+    echo "device_info{disk=\"${disk}\",type=\"${disk_type}\",model_family=\"${model_family}\",device_model=\"${device_model}\",serial_number=\"${serial_number}\",firmware_version=\"${fw_version}\"} 1"
+  elif [[ -n "${vendor}" ]] ; then
+    echo "device_info{disk=\"${disk}\",type=\"${disk_type}\",vendor=\"${vendor}\",product=\"${product}\",revision=\"${revision}\",lun_id=\"${lun_id}\"} 1"
+  fi
+  echo "device_smart_available{disk=\"${disk}\",type=\"${disk_type}\"} ${smart_available}"
+  echo "device_smart_enabled{disk=\"${disk}\",type=\"${disk_type}\"} ${smart_enabled}"
+}
+
+output_format_awk="$(cat << 'OUTPUTAWK'
+BEGIN { v = "" }
+v != $1 {
+  print "# HELP smartmon_" $1 " SMART metric " $1;
+  print "# TYPE smartmon_" $1 " gauge";
+  v = $1
+}
+{print "smartmon_" $0}
+OUTPUTAWK
+)"
+
+format_output() {
+  sort \
+  | awk -F'{' "${output_format_awk}"
+}
+
+smartctl_version="$(/usr/sbin/smartctl -V | head -n1  | awk '$1 == "smartctl" {print $2}')"
+
+echo "smartctl_version{version=\"${smartctl_version}\"} 1" | format_output
+
+if [[ "$(expr "${smartctl_version}" : '\([0-9]*\)\..*')" -lt 6 ]] ; then
+  exit
+fi
+
+device_list="$(/usr/sbin/smartctl --scan-open | awk '{print $1 "|" $3}')"
+
+for device in ${device_list}; do
+  disk="$(echo ${device} | cut -f1 -d'|')"
+  type="$(echo ${device} | cut -f2 -d'|')"
+  echo "smartctl_run{disk=\"${disk}\",type=\"${type}\"}" $(TZ=UTC date '+%s')
+  # Get the SMART information
+  /usr/sbin/smartctl -i -d "${type}" "${disk}" | parse_smartctl_info "${disk}" "${type}"
+  # Get the SMART attributes
+  /usr/sbin/smartctl -A -d "${type}" "${disk}" | parse_smartctl_attributes "${disk}" "${type}"
```

Comment:
I think it'd be useful to add a gauge with "number of 'unknown attributes' found", both not to lose information and to sense new drives not yet in smartmontools database (i.e. usually with a high number of unknown attributes, IME)